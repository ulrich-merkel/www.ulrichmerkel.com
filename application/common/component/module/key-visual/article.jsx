/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires common/component/element/headline
 * @requires common/component/element/paragraph
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 *
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Headline } from '../../element/headline';
import { P } from '../../element/paragraph';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function ModuleKeyVisualArticle(props) {
    const { headline, lead } = props;

    if (!headline || !lead) {
        return null;
    }

    const componentArticleClassName = classnames('m-key-visual__text');

    return (
        <article className={componentArticleClassName}>
            <Headline
                className="m-key-visual__headline"
                itemProp="headline"
                htmlElement="h3"
            >
                <span className="m-key-visual__boxed">{headline}</span>
            </Headline>
            <P className="m-key-visual__lead" itemProp="description">
                <span className="m-key-visual__boxed">{lead}</span>
            </P>
        </article>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [headline=''] - The article headline content
 * @property {string} [lead=''] - The article lead content
 */
ModuleKeyVisualArticle.propTypes = {
    headline: PropTypes.string,
    lead: PropTypes.string
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleKeyVisualArticle.propTypes
 */
ModuleKeyVisualArticle.defaultProps = {
    headline: '',
    lead: ''
};

export default ModuleKeyVisualArticle;
