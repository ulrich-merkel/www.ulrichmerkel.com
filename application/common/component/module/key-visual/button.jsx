/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires classnames
 *
 * @changelog
 * - 0.0.3 moved to stateless function
 * - 0.0.2 rewritten for es2015
 * - 0.0.1 basic functions and structure
 *
 */
import React, { PropTypes } from 'react';
import classnames from 'classnames';

import Button from './../../element/button';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {Object} [props] - The current component props
 * @returns {React.Element} React component markup
 */
function ModuleKeyVisualButton(props) {

    const {
        onClick,
        label,
        title
    } = props;

    if (!label || !title) {
        return null;
    }

    const componentButtonClassName = classnames(
        'm-key-visual__button--down',
        'c-font-icon--chevron-down'
    );

    return (
        <Button
            className={componentButtonClassName}
            onClick={onClick}
            title={title}
            isLabelHidden
        >
            {label}
        </Button>
    );

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {string} [headline] - The button title content
 * @property {string} [lead] - The button label content
 * @property {Function} [onClickBtn] - The button onclick handler
 */
ModuleKeyVisualButton.propTypes = {
    title: PropTypes.string,
    label: PropTypes.string,
    onClick: PropTypes.func
};

export default ModuleKeyVisualButton;
