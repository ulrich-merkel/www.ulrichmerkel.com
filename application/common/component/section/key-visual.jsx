/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires react
 * @requires component/section/common/grid
 * @requires component/module/keyVisual
 *
 * @changelog
 * - 0.0.4 added SectionCommonGridDefault
 * - 0.0.3 moved to stateless function
 * - 0.0.2 rewritten for es2015
 * - 0.0.1 basic functions and structure
 */
import React, { PropTypes } from 'react';

import SectionCommonGrid from './common/grid';
import ModuleKeyVisual from './../module/key-visual';

/**
 * Function representing a component to return a single react child element.
 *
 * @constructor
 * @param {Object} [props] The current component props
 * @returns {ReactElement} React component markup
 */
function SectionKeyVisual(props) {

    const { content, children, ...moduleProps } = props;

    return (
        <SectionCommonGrid>
            <ModuleKeyVisual
                {...{ content }}
                {...moduleProps}
            >
                {children}
            </ModuleKeyVisual>
        </SectionCommonGrid>
    );

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {Array|string} [children] The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @property {Object} [content] The content config input
 */
SectionKeyVisual.propTypes = {
    children: PropTypes.node,
    content: PropTypes.object
};

export default SectionKeyVisual;
