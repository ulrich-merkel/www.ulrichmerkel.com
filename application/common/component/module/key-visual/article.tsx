/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';

import { Headline } from '../../element/headline';
import { P } from '../../element/paragraph';
import { Article } from '../../element/article';

type Props = {
    headline?: string;
    lead?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleKeyVisualArticle: FunctionComponent<Props> = (props) => {
    const { headline, lead } = props;

    if (!headline || !lead) {
        return null;
    }

    const componentArticleClassName = classnames('m-key-visual__text');

    return (
        <Article className={componentArticleClassName}>
            <Headline
                className="m-key-visual__headline"
                itemProp="headline"
                htmlElement="h3"
            >
                <span className="m-key-visual__boxed">{headline}</span>
            </Headline>
            <P className="m-key-visual__lead" itemProp="description">
                <span className="m-key-visual__boxed">{lead}</span>
            </P>
        </Article>
    );
};
