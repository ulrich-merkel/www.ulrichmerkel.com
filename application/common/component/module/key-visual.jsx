/* eslint-disable immutable/no-mutation, immutable/no-this */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { connect } from 'react-redux';

import { selectStateReducedMotionSelected } from '../../state/reduced-motion/selector';
import { scrollTo } from '../../utils/scroll-to';
import { ModuleKeyVisualPicture } from './key-visual/picture';
import { ModuleKeyVisualArticle } from './key-visual/article';
import { ModuleKeyVisualButton } from './key-visual/button';
import { eventPreventDefault } from '../../utils/event';

/**
 * Class representing a component.
 *
 * @class
 * @augments React.Component
 */
export class ModuleKeyVisual extends Component {
    /**
     * Handle button down click event.
     *
     * @param {object} event - Click event object
     * @returns {void}
     */
    onClickBtn = (event) => {
        const { onClickBtn, reducedMotionSelected } = this.props;
        const { keyVisual } = this;

        eventPreventDefault(event);

        if (keyVisual) {
            const innerHeight = keyVisual.clientHeight;
            const boundingClientRectTop = Math.abs(
                keyVisual.getBoundingClientRect().top
            );
            const offsetTop = document.body.scrollTop;

            scrollTo({
                duration: reducedMotionSelected ? 0 : 300,
                top: innerHeight + boundingClientRectTop + offsetTop
            });
            onClickBtn();
        }
    };

    /**
     * The required render function to return a single react child element.
     *
     * @returns {ReactElement} React component markup
     */
    render() {
        const {
            componentType,
            className,
            isWork,
            isCovered,
            content,
            ...otherProps
        } = this.props;

        if (!content.img) {
            return null;
        }

        const ComponentType = componentType;
        const componentClassName = classnames(
            isWork ? 'm-key-visual--work' : 'm-key-visual',
            className
        );

        return (
            <ComponentType
                className={componentClassName}
                ref={(ref) => {
                    this.keyVisual = ref;
                }}
                {...otherProps}
            >
                <ModuleKeyVisualPicture
                    img={content.img}
                    type={content.type}
                    isCovered={isCovered}
                />
                <ModuleKeyVisualArticle
                    headline={content.headline}
                    lead={content.lead}
                />
                <ModuleKeyVisualButton
                    onClick={this.onClickBtn}
                    label={content.btnLabel}
                    title={content.btnTitle}
                />
            </ComponentType>
        );
    }
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [componentType='div'] - The component element type used for React.createElement
 * @property {string} [className] - The component css class names - will be merged into component default classNames
 * @property {boolean} [isWork=false] - Whether the component has work modifiers or not
 * @property {boolean} [isCovered=false] - Whether the component should be background size covered or not
 * @property {object} [content={}] - The component translation config
 */
ModuleKeyVisual.propTypes = {
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    componentType: PropTypes.string,
    content: PropTypes.shape({
        headline: PropTypes.string,
        lead: PropTypes.string,
        btnLabel: PropTypes.string,
        btnTitle: PropTypes.string,
        img: PropTypes.object,
        type: PropTypes.string
    }),
    isCovered: PropTypes.bool,
    isWork: PropTypes.bool,
    onClickBtn: PropTypes.func,
    reducedMotionSelected: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleKeyVisual.propTypes
 */
ModuleKeyVisual.defaultProps = {
    componentType: 'div',
    content: {},
    isCovered: false,
    isWork: false,
    onClickBtn: Function.prototype,
    reducedMotionSelected: false
};

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the componentâ€™s props.
 *
 * @private
 * @param {object<string, *>} state - The current redux store state
 * @returns {object<string, *>} The mapped state properties
 */
function mapStateToProps(state) {
    return {
        reducedMotionSelected: selectStateReducedMotionSelected(state)
    };
}

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 */
export const ModuleKeyVisualConnected = connect(mapStateToProps)(
    ModuleKeyVisual
);
