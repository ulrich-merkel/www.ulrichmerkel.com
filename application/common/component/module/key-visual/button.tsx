/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';
import { noop } from 'lodash';

import { Button } from '../../element/button';

type Props = {
    label?: string;
    onClick?: () => void;
    title?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleKeyVisualButton: FunctionComponent<Props> = (props) => {
    const { onClick = noop, label, title } = props;

    if (!label || !title) {
        return null;
    }

    const componentButtonClassName = classnames(
        'm-key-visual__button--down',
        'c-font-icon--chevron-down'
    );

    return (
        <Button
            className={componentButtonClassName}
            onClick={onClick}
            title={title}
            isLabelHidden
        >
            {label}
        </Button>
    );
}
