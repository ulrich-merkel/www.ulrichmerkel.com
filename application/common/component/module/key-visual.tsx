/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, createRef, Component } from 'react';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { noop } from 'lodash';

import { selectStateReducedMotionSelectedReduce } from '../../state/reduced-motion/selector';
import { scrollTo } from '../../utils/scroll-to';
import { ModuleKeyVisualPicture } from './key-visual/picture';
import { ModuleKeyVisualArticle } from './key-visual/article';
import { ModuleKeyVisualButton } from './key-visual/button';
import { eventPreventDefault } from '../../utils/event';

type Props = {
    className?: string;
    content?: {
        headline: string;
        lead: string;
        btnLabel: string;
        btnTitle: string;
        img: any;
        type: string;
    };
    htmlElement?: string;
    isCovered?: boolean;
    isWork?: boolean;
    onClickBtn?: () => void;
    reducedMotionSelected?: boolean;
}

/**
 * Class representing a component.
 *
 * @class
 * @augments React.Component
 */
export class ModuleKeyVisual extends Component<Props> {
    private keyVisual = createRef<HTMLDivElement>();

    /**
     * Handle button down click event.
     *
     * @param {object} event - Click event object
     * @returns {void}
     */
    onClickBtn = (event) => {
        const { onClickBtn = noop, reducedMotionSelected = false } = this.props;

        eventPreventDefault(event);

        if (this.keyVisual) {
            const innerHeight = this.keyVisual.current.clientHeight;
            const boundingClientRectTop = Math.abs(
                this.keyVisual.current.getBoundingClientRect().top
            );
            const offsetTop = document.body.scrollTop;

            scrollTo({
                duration: reducedMotionSelected ? 0 : 300,
                top: innerHeight + boundingClientRectTop + offsetTop
            });
            onClickBtn();
        }
    };

    /**
     * The required render function to return a single react child element.
     *
     * @returns {ReactElement} React component markup
     */
    render() {
        const {
            className,
            content,
            htmlElement: HtmlElement = 'div',
            isCovered = false,
            isWork = false,
            ...otherProps
        } = this.props;

        if (!content.img) {
            return null;
        }

        const componentClassName = classnames(
            isWork ? 'm-key-visual--work' : 'm-key-visual',
            className
        );

        return (
            <HtmlElement
                className={componentClassName}
                ref={this.keyVisual}
                {...otherProps}
            >
                <ModuleKeyVisualPicture
                    img={content.img}
                    type={content.type}
                    isCovered={isCovered}
                />
                <ModuleKeyVisualArticle
                    headline={content.headline}
                    lead={content.lead}
                />
                <ModuleKeyVisualButton
                    onClick={this.onClickBtn}
                    label={content.btnLabel}
                    title={content.btnTitle}
                />
            </HtmlElement>
        );
    }
}

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the componentâ€™s props.
 *
 * @private
 * @param {object<string, *>} state - The current redux store state
 * @returns {object<string, *>} The mapped state properties
 */
function mapStateToProps(state) {
    return {
        reducedMotionSelected: selectStateReducedMotionSelectedReduce(state)
    };
}

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {ReactElement}
 */
export const ModuleKeyVisualConnected = connect(mapStateToProps)(
    ModuleKeyVisual
);
