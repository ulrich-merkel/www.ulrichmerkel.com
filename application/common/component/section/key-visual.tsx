/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';

import { SectionCommonGrid } from './common/grid';
import { ModuleKeyVisualConnected } from '../module/key-visual';

type Props = {
    children?: ReactNode;
    content: any;
    className?: string;
    isCovered?: boolean;
    isWork?: boolean;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @param {Array|string} [props.children] - The component dom node childs, usally an array of components, if there is only a single child it's a string
 * @param {object} [props.content={}] - The content config input
 * @returns {ReactElement} React component markup
 */
export const SectionKeyVisual: FunctionComponent<Props> = (props) => {
    const { content, children, className, isCovered, isWork } = props;

    return (
        <SectionCommonGrid>
            <ModuleKeyVisualConnected {...{ content, className, isCovered, isWork }}>
                {children}
            </ModuleKeyVisualConnected>
        </SectionCommonGrid>
    );
}
