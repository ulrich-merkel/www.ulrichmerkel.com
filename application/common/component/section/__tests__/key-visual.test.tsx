import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { SectionKeyVisual } from '../key-visual';

describe('SectionKeyVisual', function fnDescribe() {
    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <SectionKeyVisual content={{}}>
                Section key-visual children
            </SectionKeyVisual>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
