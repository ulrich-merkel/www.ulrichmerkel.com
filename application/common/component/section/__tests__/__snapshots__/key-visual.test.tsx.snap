// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionKeyVisual should render correctly 1`] = `
<DocumentFragment>
  <section
    class="l-grid l-main__transition"
  >
    <div
      class="l-grid__row"
    >
      <div
        class="l-grid__col--12"
      />
    </div>
  </section>
</DocumentFragment>
`;
